File: solutions/TypeScript/[0202_Easy] Happy Number/isHappy.ts
function isHappy(n: number): boolean {
  const hashMap: Record<number, number> = {};
  let squaresSum = n;

  while (squaresSum !== 1) {
    let temp = 0;
    for (let i = squaresSum; i > 0; i = Math.floor(i / 10)) {
      temp += Math.pow(i % 10, 2);
    }
    squaresSum = temp;

    if (hashMap[squaresSum]) {
      return false;
    }

    hashMap[squaresSum] = 1;
  }

  return true;
}

File: solutions/Java/[0202_Easy] Happy Number/isHappy.java
class Solution {
  public boolean isHappy(int n) {
    HashMap<Integer, Integer> map = new HashMap<>();
    int squaresSum = n;

    while (squaresSum != 1) {
      int temp = 0;
      for (int i = squaresSum; i > 0; i /= 10) {
        temp += Math.pow(i % 10, 2);
      }
      squaresSum = temp;

      if (map.containsKey(squaresSum)) {
        return false;
      }

      map.put(squaresSum, 1);
    }

    return true;
  }
}
